CoinList 화면 콜렉션 뷰의 데이터 소스 클래스의 함수를 구현

z3rosmith opened the pull request:
- CoinList화면에 통신을 적용했습니다
- 코인 symbol에 대한 한글이름을 출력하기 위한 Localized.strings 파일 추가
- coinList를 sort할 떄 CoinSortType(enum)을 전달해주는 방식에서 CoinSortAction(closure)을 전달해주는 방식으로 바꿈
- Coin 타입을 CoinListDataManager에서 분리
- 현재가 불러오는 통신 구현

z3rosmith committed:
feat: CoinList 네트워킹 적용
- CoinList화면에 통신을 적용했습니다
- CoinSortType에 따른 정렬을 적용했습니다

z3rosmith committed:
feat: Localized.strings 파일 추가
- 코인 symbol에 대한 한글이름을 출력하기 위한 Localized.strings 파일
  추가
- coinList를 sort할 떄 CoinSortType(enum)을 전달해주는 방식에서
  CoinSortAction(closure)을 전달해주는 방식으로 바꿈

z3rosmith committed:
feat: coinList화면에서 현재가 불러오는 통신 구현
- Coin 타입을 CoinListDataManager에서 분리
- 현재가 불러오는 통신 구현

z3rosmith committed:
refactor: Coin타입 initializer 삭제 #26
- Coin타입 initializer 삭제

HJEHA commented on BithumbYagomAcamedy/Controller/CoinListDataManager.swift:
```diff
@@ -0,0 +1,154 @@
+// MARK: - Networking
+
+extension CoinListDataManager {
+    #warning("response status else 구문 구현 필요")
```
`#warning`를 남겨두신 이유가 있을까요???:)

z3rosmith replied:
추후에 꼭 추가되어야 할것같은데 잊을 수도 있을거같아 `#warning` 으로 남겨두었습니다. 아니면 따로 문서로 남겨두는게 더 나은 방법일까요?

z3rosmith replied:
아래에 동건이 남겨주신 댓글처럼 Todo 주석으로 반영해보도록 하겠습니다!

HJEHA commented on BithumbYagomAcamedy/Controller/CoinListDataManager.swift:
```diff
@@ -0,0 +1,154 @@
+                } catch {
+                    print(error)
+                }
+            case .failure(let error):
+                if let description = error.errorDescription {
+                    print(description)
+                }
+            }
```
print(error.localizedDescription)으로 출력해도 괜찮을 것 같습니다:)

donggeonoh replied:
위에 catch 문 에서도요 ㅎㅎ

HJEHA commented on BithumbYagomAcamedy/Model/ValueObject/TickersValueObject.swift:
```diff
@@ -70,8 +70,22 @@ struct TickerData: Decodable {
+extension TickerData {
+    var accTradeValue24HourDouble: Double {
+        return Double(accTradeValue24Hour) ?? -1 // 의견 필요
+    }
```
어떤 상황인지 회의실에서 이야기 해주실 수 있으신가요???

HJEHA replied:
### 회의 내용
- 변동률이나 현재가에 -1또는 0 등 숫자를 반환할 경우 사용자가 수치를 오해할 수 있는 문제가 발생할 것 같다.
- 안전하게 옵셔널로 처리하는 것이 맞다고 판단된다.
- 옵셔널이 들어올 경우 cellView에서 값이 옵셔널일 경우 "데이터를 가져오는데 실패했습니다" 같은 메세지를 해당 `Label`에 출력해주는 방식으로 개선해 볼 수 있을 것 같다.

HJEHA commented on BithumbYagomAcamedy/Controller/CoinListViewController.swift:
```diff
@@ -27,31 +27,70 @@ class CoinListViewController: UIViewController {
-    private var dataSource: UICollectionViewDiffableDataSource<Section, CoinListDataSource.Coin>?
+    private let coinListDataManager = CoinListDataManager()
+    private var dataSource: UICollectionViewDiffableDataSource<Section, Coin>?
+    private var coinSortAction: CoinSortAction? {
+        didSet {
+            guard let coinSortAction = coinSortAction else { return }
+            applySnapshot(by: coinSortAction)
+        }
+    }
```
`coinSortAction`은 `CoinListDataManage`가 가지고 있는게 VC의 역할을 최소화하는데 도움이 될것 같아요:)
정렬 버튼 `@IBAction`이 눌렸을 때  `CoinListDataManage`에게 `coinSortAction`을 넘겨주고 
`CoinListDataManage`는 코인을 정렬 시킨 후 `delegate(VC)`에게 `applySnapshot(정렬된 코인 리스트)`를 수행하는건 어떠신가요???

donggeonoh replied:
허황 의견에 동의합니다~ 저도 일단 VC에서 처리하였는데, Model에서 처리하도록 구현중입니다.
그렇게 해야 저희가 원했던 VC에 거대해지는 현상이 없을 것 같습니다~

z3rosmith replied:
좋은것같습니다! 반영하겠습니다.

HJEHA reviewed (commented):
코멘트 확인 부탁드립니다:)

donggeonoh commented on BithumbYagomAcamedy/Controller/CoinListDataManager.swift:
```diff
@@ -0,0 +1,154 @@
+
+typealias CoinSortAction = (Coin, Coin) -> Bool
+
+protocol CoinListDataManagerDelegate {
+    func coinListDataManagerDidSetCoinList()
+    func coinListDataManagerDidFetchCurrentPrice()
+}
```
DidSet과 DidFetch로 네이밍을 다르게 가져가신 이유가 뭘까요?

z3rosmith replied:
DidSet은 coinlist의 값을 sett해주었을때 불리는 메소드이고
DidFetch는 current price를 네트워킹을 통해 fetch해왔을때 불리는 메소드입니다.

donggeonoh replied:
아 그렇군요 알겠습니다 ㅎㅎ

donggeonoh commented on BithumbYagomAcamedy/Controller/CoinListDataManager.swift:
```diff
@@ -0,0 +1,154 @@
+
+// MARK: - Data Processing
+
+extension CoinListDataManager {
+    func sortedCoinList(by areInIncreasingOrder: @escaping CoinSortAction) -> [Coin] {
+        return coinList.sorted(by: areInIncreasingOrder)
+    }
+}
```
parameter name은 increasingOrder 어때요? by 인자 라벨이 붙어서 매끄럽지 못한 것 같아요

z3rosmith replied:
Swift의 Array의 sorted메소드 네이밍을 그대로 가져왔습니다. 의미상으로 봐도 `areInIncreasingOrder`가 CoinSortAction이 어떤 기준을 잡아주는것에대해서인지 더 명확하게 알려주는것같습니다

donggeonoh replied:
아 그렇군요~~~ 감사합니다 ㅎㅎ

donggeonoh commented on BithumbYagomAcamedy/Controller/CoinListDataManager.swift:
```diff
@@ -0,0 +1,154 @@
+
+import Foundation
+
+typealias CoinSortAction = (Coin, Coin) -> Bool
```
CoinListDataManager 에서만 쓰인다면 클래스 안에 넣어도 괜찮다고 생각합니다.

z3rosmith replied:
지금은 CoinListViewController에서도 사용하고있습니다. 다만 허황의견대로 sortAction을 뷰컨트롤러에서 가지고있는게 좋지 않다고 판단되어 수정할 때 반영하겠습니다.

donggeonoh commented on BithumbYagomAcamedy/Controller/CoinListDataManager.swift:
```diff
@@ -0,0 +1,154 @@
+                    let response = try JSONParser().decode(data: data, type: TickersValueObject.self)
+                    if response.status == "0000" {
+                        self?.setCoinList(from: response)
+                        #warning("밑에줄 주석처리하면 괜찮음.. 왜??")
```
저희가 develop에 반영할 코드라 warning은 빼주셨으면 좋겠습니다. 메서드에 Todo로 주석을 남겨보는건 어떠신가요?

z3rosmith replied:
좋은것같습니다 Todo 주석으로 바꾸겠습니다!

donggeonoh commented on BithumbYagomAcamedy/Controller/CoinListDataManager.swift:
```diff
@@ -0,0 +1,154 @@
+            case .success(let data):
+                do {
+                    let response = try JSONParser().decode(data: data, type: TickersValueObject.self)
+                    if response.status == "0000" {
```
개인적인 의견으로 0000 문자열을 계속 사용한다면 클래스 private로 하여 프로퍼티로 가지고 있어도 괜찮다고 생각이 드네요

z3rosmith replied:
넵 말씀대로 클래스 privaet메서드로 가지는것으로 수정해보겠습니다

donggeonoh commented on BithumbYagomAcamedy/Controller/CoinListDataManager.swift:
```diff
@@ -0,0 +1,154 @@
+                    if response.status == "0000" {
+                        self?.setCoinList(from: response)
+                        #warning("밑에줄 주석처리하면 괜찮음.. 왜??")
+//                        self?.delegate?.coinListDataManagerDidSetCoinList()
```
주석은 삭제해주세요 ㅎㅎ

donggeonoh commented on BithumbYagomAcamedy/Controller/CoinListDataManager.swift:
```diff
@@ -0,0 +1,154 @@
+                        self?.fetchCurrentPrice()
+                    }
+                } catch {
+                    print(error)
```
에러 출력 할 때  error.localizedescription 으로 출력하면 더 좋을 것 같습니다.

donggeonoh commented on BithumbYagomAcamedy/Controller/CoinListDataManager.swift:
```diff
@@ -0,0 +1,154 @@
+        }
+    }
+    
+    private func fetchCurrentPrice() {
+        var count: Int = 0
+        let serialQueue = DispatchQueue(label: "serial")
+        for i in 0..<coinList.count {
```
let 과 for문 줄바꿈 해주세요 ㅎㅎ

donggeonoh commented on BithumbYagomAcamedy/Controller/CoinListDataManager.swift:
```diff
@@ -0,0 +1,154 @@
+                case .success(let data):
+                    do {
+                        let response = try JSONParser().decode(data: data, type: TranscationValueObject.self)
+                        if response.status == "0000" {
```
guard를 활용한다면 depth는 일단 줄어들 것 같네요 ㅎㅎ

z3rosmith replied:
좋은것같습니다.

donggeonoh commented on BithumbYagomAcamedy/Controller/CoinListDataManager.swift:
```diff
@@ -0,0 +1,154 @@
+                                }
+                            }
+                        }
+                    } catch {
+                        print(error)
+                    }
+                case .failure(let error):
+                    if let description = error.errorDescription {
+                        print(description)
+                    }
+                }
```
error를 error.localized.... 를 print 하시면 더 좋을 것 같습니다 ㅎㅎ

donggeonoh commented on BithumbYagomAcamedy/Controller/CoinListViewController.swift:
```diff
@@ -27,31 +27,70 @@ class CoinListViewController: UIViewController {
+        coinListDataManager.fetchCoinList()
+    }
+    
+    @IBAction func nameButtonTapped(_ sender: UIButton) {
+        if sender.isSelected {
+            coinSortAction = { $0.callingName > $1.callingName }
+        } else {
+            coinSortAction = { $0.callingName < $1.callingName }
+        }
+        sender.isSelected.toggle()
+    }
```
깔끔해요!!! 굿굿

donggeonoh commented on BithumbYagomAcamedy/Controller/CoinListViewController.swift:
```diff
@@ -80,25 +119,33 @@ extension CoinListViewController {
+        let allCoinList = coinListDataManager.sortedCoinList(by: areInIncreasingOrder)
+        var snapshot = NSDiffableDataSourceSnapshot<Section, Coin>()
+        snapshot.appendSections([.allByKRW])
+        snapshot.appendItems(allCoinList, toSection: .allByKRW)
```
Section 이름을 그냥 all로 가져가는게 어떠실까요?

z3rosmith replied:
all을 나눌 필요는 없으니 all로 가져가는게 좋을것같네요 의견 감사합니다:)

donggeonoh commented on BithumbYagomAcamedy/Model/ValueObject/TickersValueObject.swift:
```diff
@@ -70,8 +70,22 @@ struct TickerData: Decodable {
+        return Double(accTradeValue24Hour) ?? -1 // 의견 필요
+    }
+    
+    var fluctate24HourDouble: Double {
+        return Double(fluctate24Hour) ?? -1 // 의견 필요
+    }
     
+    var fluctateRate24HourDouble: Double {
+        return Double(fluctateRate24Hour) ?? -1.0 // 의견 필요
+    }
```
전에 얘기했던대로 옵셔널으로 하는 것이 좋아보입니다.

donggeonoh commented on BithumbYagomAcamedy/Model/ValueObject/TransactionValueObject.swift:
```diff
@@ -24,3 +24,10 @@ struct TransactionData: Decodable {
     let total: String
 }
+
+#warning("의견필요 - 기본값 처리")
+extension TransactionData {
+    var priceDouble: Double {
+        return Double(price) ?? -1
+    }
+}
```
이부분도 옵셔널로 처리해도 문제 없을까요?

z3rosmith replied:
이부분도 옵셔널로 처리하겠습니다!

donggeonoh commented on BithumbYagomAcamedy/View/CoinListView/CoinListCollectionViewCell.swift:
```diff
@@ -9,17 +9,17 @@ import UIKit
 
 class CoinListCollectionViewCell: UICollectionViewListCell {
```
final 까지 붙으면 더 좋겠네요 ㅎㅎ

z3rosmith committed:
refactor: CoinListDataSource PR사항 반영 #26
- coinSortAction 프로퍼티를 ViewController에서 DataManager로 옮겼습니다
- Coin의 가격 관련 프로퍼티들을 옵셔널로 바꾸었습니다
- 기타 지적사항 반영하였습니다

z3rosmith committed:
refactor: Coin의 popularity 프로퍼티 let에서 var로 변경 #26
- 제목이 곧 내용

z3rosmith committed:
refactor: final keyword 추가
- CoinListCollectionViewCell과 CoinListViewController에 final 키워드
  추가

z3rosmith committed:
Merge branch 'develop' of https://github.com/DonggeonOh/bithumb-techcamp-ios-1st into feat/26/implement-coinlistdatasource-functions
